supernode/node/grpc/server/services/supernode/process_userdata.go

```diff
@@ -0,0 +1,176 @@
+	"strings"
+
+	"github.com/pastelnetwork/gonode/common/service/userdata"
+
+	"github.com/pastelnetwork/gonode/common/errors"
+	"github.com/pastelnetwork/gonode/common/log"
+	"github.com/pastelnetwork/gonode/metadb/database"
+	pb "github.com/pastelnetwork/gonode/proto/supernode"
+	"github.com/pastelnetwork/gonode/supernode/node/grpc/server/services/common"
+	"github.com/pastelnetwork/gonode/supernode/services/userdataprocess"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/peer"
+)
+
+const defaultGRPCPort = 4444
+
+// ProcessUserdata represents grpc service for processing userdata.
+type ProcessUserdata struct {
+	pb.UnimplementedProcessUserdataServer
+
+	*common.ProcessUserdata
+	databaseOps *database.Ops
+}
+
+// Session implements supernode.ProcessUserdataServer.Session()
+func (service *ProcessUserdata) Session(stream pb.ProcessUserdata_SessionServer) error {
+	ctx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+
+	var task *userdataprocess.Task
+
+	if sessID, ok := service.SessID(ctx); ok {
+		task = service.Task(sessID)
+		if task == nil {
+			return errors.Errorf("not found %q task", sessID)
+		}
+	} else {
+		task = service.NewTask()
+	}
+	go func() {
+		<-task.Done()
+		cancel()
+	}()
+	defer task.Cancel()
+
+	peer, _ := peer.FromContext(ctx)
+	log.WithContext(ctx).WithField("addr", peer.Addr).Debugf("Session stream")
+	defer log.WithContext(ctx).WithField("addr", peer.Addr).Debugf("Session stream closed")
+
+	req, err := stream.Recv()
+	if err != nil {
+		return errors.Errorf("failed to receieve handshake request: %w", err)
+	}
+	log.WithContext(ctx).WithField("req", req).Debugf("Session request")
+
+	if err := task.SessionNode(ctx, req.NodeID); err != nil {
+		return err
+	}
+
+	resp := &pb.MDLSessionReply{
+		SessID: task.ID(),
+	}
+	if err := stream.Send(resp); err != nil {
+		return errors.Errorf("failed to send handshake response: %w", err)
+	}
+	log.WithContext(ctx).WithField("resp", resp).Debugf("Session response")
+
+	for {
+		if _, err := stream.Recv(); err != nil {
+			if err == io.EOF {
+				return errors.Errorf("handshake stream closed: %w", err)
+			}
+		}
+	}
+}
+
+// SendUserdataToPrimary implements supernode.ProcessUserdataServer.SendUserdataToPrimary()
+func (service *ProcessUserdata) SendUserdataToPrimary(ctx context.Context, req *pb.SuperNodeRequest) (*pb.SuperNodeReply, error) {
+	// This code run in primary supernode
+	log.WithContext(ctx).WithField("req", req).Debugf("SendUserdataToPrimary request")
+
+	if req == nil {
+		return nil, errors.Errorf("receive nil request")
+	}
+
+	task, err := service.TaskFromMD(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Primary node will connect to leader node here.
+	if !service.databaseOps.IsLeader() {
+		var extAddress string
+
+		leaderAddress := service.databaseOps.LeaderAddress()
+		parts := strings.Split(leaderAddress, ":")
+		//temporary use this to test on local machine
+		localEnv := false
+		if localEnv {
+			port := parts[len(parts)-1]
```

Comment:
should validate `len(parts)` first

Reply:
len of parts cannot be 0, min value is 1. So we just expect the last element of parts is the port value
thanks